src/main/kotlin/cacophony/parser/LLOneParser.kt

```diff
@@ -1,15 +1,79 @@
+            val nextAction: Map<SymbolType, MutableMap<DFAStateReference<StateType, SymbolType, Production<SymbolType>>, SymbolType?>> =
+                enumValues<SymbolType>().toList().associateWith { mutableMapOf() }
+
+            analyzedGrammar.automata.forEach { (dfaLabel, dfa) -> // automaton for the grammar symbol
+                dfa.getAllStates().forEach { curState -> // dfa state we are currently in
+                    val curStateRef = DFAStateReference(curState, dfa)
+
+                    enumValues<SymbolType>().forEach { inputSymbol -> // symbol from the input
+                        val suitableSymbols = mutableListOf<SymbolType>()
+                        enumValues<SymbolType>().forEach prod@{ prodSymbol -> // symbol in the production
+                            dfa.getProduction(curState, prodSymbol) ?: return@prod
+
+                            val dfaForProdSymbol = analyzedGrammar.automata[prodSymbol]
+                            if (dfaForProdSymbol != null) { // nonterminal
+                                val nextState = dfaForProdSymbol.getStartingState()
+                                val nextStateRef = DFAStateReference(nextState, dfaForProdSymbol)
+                                if ((analyzedGrammar.first[nextStateRef]?.contains(inputSymbol) == true) or (prodSymbol == inputSymbol)) {
+                                    suitableSymbols.add(prodSymbol)
+                                } else if (analyzedGrammar.nullable.contains(nextStateRef) and
+                                    (analyzedGrammar.follow[nextStateRef]?.contains(inputSymbol) == true)
+                                ) {
+                                    suitableSymbols.add(prodSymbol)
+                                }
+                            } else { // terminal
+                                if (prodSymbol == inputSymbol) {
+                                    suitableSymbols.add(prodSymbol)
+                                }
+                            }
+                        }
+
+                        // Leaving this for the convenience in (very) possible future debugging.
+//                        println("state $curState in $dfaLabel, possibilities for $inputSymbol are $suitableSymbols")
+
+                        if (suitableSymbols.size > 1) { // too many productions are suitable
+                            throw ParserConstructorErrorException(
+                                "Not an LL(1) grammar: " +
+                                    "state $curState in dfa $dfaLabel, possible productions for $inputSymbol are $suitableSymbols",
+                            )
+                        }
+                        if (suitableSymbols.size == 1) {
+                            nextAction[inputSymbol]!![curStateRef] = suitableSymbols[0]
+                        } else { // acceptableSymbols.size == 0
+                            nextAction[inputSymbol]!![curStateRef] = null
+                        }
```

Comment:
this is just an aesthetic comment, but this can be rewritten as a really neat when statement:

```kotlin
when (suitableSymbols.size) {
    0 -> nextAction[inputSymbol]!![curStateRef] = null
    1 -> nextAction[inputSymbol]!![curStateRef] = suitableSymbols[0]
    else -> throw ParserConstructorErrorException(
        "Not an LL(1) grammar: " + 
        "state $curState in dfa $dfaLabel, possible productions for $inputSymbol are $suitableSymbols",
    )
}
```